src/julia.h

```diff
@@ -278,57 +278,80 @@ typedef jl_value_t *(*jl_fptr_sparam_t)(jl_svec_t*, jl_value_t*, jl_value_t**, u
-                                // used to avoid infinite recursion
-    uint8_t inCompile : 1;
-    jl_fptr_t fptr;             // jlcall entry point
+    // cache of specializations of this method for invoke(), i.e.
+    // cases where this method was called even though it was not necessarily
+    // the most specific for the argument types.
+    struct _jl_methtable_t *invokes;
+    struct _jl_method_t *next;
+} jl_method_t;
+
+typedef struct _jl_ast_info_t {
+    JL_DATA_TYPE
+    // this holds the static data for a (possibly inferred) syntax tree
+    jl_value_t *ast;
+    jl_method_t *def;  // method this is specialized from
+    // a slower-but-works version of this function as a fallback
+    // this keeps codegen happy if the method contains intrinsics that depend on an sparam
+    struct _jl_lambda_info_t *unspecialized_ducttape;
+    uint32_t called;  // bit flags: whether each of the first 8 arguments is called
+    uint8_t pure;
+    uint8_t inferred;
+    uint8_t need_ducttape; // whether this function has sparams and intrinsics,
+        // signaling that it probably can't be compiled via unspecialized
+        // (until the interpreter / runtime-intrinsics are fully functional)
+} jl_ast_info_t;
 
+typedef struct _jl_llvm_functions_t {
     // On the old JIT, handles to all Functions generated for this linfo
     // For the new JITs, handles to declarations in the shadow module
     // with the same name as the generated functions for this linfo, suitable
     // for referencing in LLVM IR
-    jl_llvm_functions_t functionObjects;
+    void *functionObject;     // jlcall llvm Function
+    void *cFunctionList;      // c callable llvm Functions
+    void *specFunctionObject; // specialized llvm Function (common core for the other two)
+    int32_t functionID;       // index that this function will have in the codegen table
+    int32_t specFunctionID;   // index that this specFunction will have in the codegen table
+    jl_fptr_t fptr;           // jlcall entry point
+    uint8_t jlcall_api;       // the c-abi for fptr; 0 = jl_fptr_t, 1 = jl_fptr_sparam_t
+} jl_llvm_functions_t;
 
-    int32_t functionID; // index that this function will have in the codegen table
-    int32_t specFunctionID; // index that this specFunction will have in the codegen table
+typedef struct _jl_lambda_info_t {
+    JL_DATA_TYPE
+    // this holds the data for a executable thunk that maps sig -> rettype
+    struct _jl_lambda_info_t *next;
+    jl_tupletype_t *sig; // aka specTypes
+    jl_value_t *rettype;
+    uint8_t va; // sig has Vararg
+    uint8_t inCompile; // flags to tell if inference is running on this function
+                       // used to avoid infinite recursion
+    // sparam_vals is the closure data (a vector of values indexed by func->def->sparam_syms)
+    jl_svec_t *sparam_vals; // NULL if this is a guard entry
+    jl_value_t *func; // where this function came from, NULL if there isn't an associated definition
```

Comment:
this commit was an exploration in making LambdaInfo the minimal set of fields to make it a purely executable thunk (without any extraneous metadata like an AST, name, debug info, etc.)

it was an abstract representation of the contract `fptr(this, sig) -> rettype` where `fptr = compile(func)`